File: notification-service/src/test/java/com/eirsteir/coffeewithme/notification/service/NotificationServiceImplTest.java
package com.eirsteir.coffeewithme.notification.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.mockito.Mockito.when;

import com.eirsteir.coffeewithme.commons.domain.notification.NotificationType;
import com.eirsteir.coffeewithme.commons.exception.CWMException;
import com.eirsteir.coffeewithme.notification.config.ModelMapperConfig;
import com.eirsteir.coffeewithme.notification.domain.Notification;
import com.eirsteir.coffeewithme.notification.dto.NotificationDto;
import com.eirsteir.coffeewithme.notification.repository.NotificationRepository;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mockito;
import org.modelmapper.ModelMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.Import;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit.jupiter.SpringExtension;

@Import({ModelMapperConfig.class})
@ExtendWith(SpringExtension.class)
@ContextConfiguration(classes = {NotificationServiceImpl.class})
class NotificationServiceImplTest {

  private final long TO_USER_ID = 1L;
  private final long CURRENT_USER_ID = 2L;

  private Notification friendRequestNotification;
  private Notification.UserDetails currentUser;

  @Autowired private NotificationService service;

  @MockBean private NotificationRepository notificationRepository;

  @Autowired private ModelMapper modelMapper;

  @BeforeEach
  void setUp() {
    currentUser = Notification.UserDetails.builder().id(CURRENT_USER_ID).name("From user").build();

    friendRequestNotification =
        Notification.builder()
            .notificationId(1L)
            .subjectId(TO_USER_ID)
            .type(NotificationType.FRIEND_REQUEST)
            .user(currentUser)
            .build();
  }

  @Test
  void testFindAllByUserWhenUserHasNotifications_thenReturnListOfNotificationDto() {
    Pageable firstPage = PageRequest.of(0, 2);
    when(notificationRepository.findAllByUser_idOrderByTimestamp(TO_USER_ID, firstPage))
        .thenReturn(
            Arrays.asList(
                Notification.builder().user(currentUser).build(),
                Notification.builder().user(currentUser).build()));

    List<NotificationDto> notifications = service.findAllByUserId(TO_USER_ID, firstPage);

    assertThat(notifications).hasSize(2);
    assertThat(notifications.get(0).getUser().getId()).isEqualTo(currentUser.getId());
  }

  @Test
  void testUpdateNotificationToReadWhenFound_thenSetIsReadToTrue() {
    friendRequestNotification.setSeen(true);

    Notification notification =
        Notification.builder()
            .notificationId(friendRequestNotification.getNotificationId())
            .user(friendRequestNotification.getUser())
            .seen(false)
            .build();

    when(notificationRepository.findById(friendRequestNotification.getNotificationId()))
        .thenReturn(Optional.ofNullable(notification));
    when(notificationRepository.save(Mockito.any(Notification.class)))
        .thenReturn(friendRequestNotification);

    NotificationDto notificationDtoToUpdate = modelMapper.map(notification, NotificationDto.class);
    NotificationDto updatedNotificationDto =
        service.updateNotificationToRead(notificationDtoToUpdate);

    assertThat(updatedNotificationDto.getSeen()).isTrue();
  }

  @Disabled
  @Test
  void testUpdateNotificationToReadWhenNotFound_thenThrowEntityNotFoundException() {
    when(notificationRepository.findById(friendRequestNotification.getNotificationId()))
        .thenReturn(Optional.empty());

    NotificationDto notificationDtoToUpdate =
        modelMapper.map(friendRequestNotification, NotificationDto.class);

    assertThatExceptionOfType(CWMException.EntityNotFoundException.class)
        .isThrownBy(() -> service.updateNotificationToRead(notificationDtoToUpdate))
        .withMessage(
            "Requested notification with id - "
                + friendRequestNotification.getNotificationId()
                + " does not exist");
  }
}

File: notification-service/src/test/kotlin/com/eirsteir/coffeewithme/notification/service/NotificationServiceImplTest.kt
package com.eirsteir.coffeewithme.notification.service

import com.eirsteir.coffeewithme.commons.domain.notification.NotificationType
import com.eirsteir.coffeewithme.commons.exception.APIException
import com.eirsteir.coffeewithme.commons.test.BaseUnitTest
import com.eirsteir.coffeewithme.notification.config.ModelMapperConfig
import com.eirsteir.coffeewithme.notification.domain.Notification
import com.eirsteir.coffeewithme.notification.dto.NotificationDto
import com.eirsteir.coffeewithme.notification.repository.NotificationRepository
import org.assertj.core.api.Assertions
import org.assertj.core.api.ThrowableAssert.ThrowingCallable
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension
import org.modelmapper.ModelMapper
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.context.annotation.Import
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.junit.jupiter.SpringExtension
import java.util.*

@ActiveProfiles("test")
@ExtendWith(MockitoExtension::class)
internal class NotificationServiceImplTest  {

    private val TO_USER_ID = 1L
    private val CURRENT_USER_ID = 2L
    private lateinit var friendRequestNotification: Notification
    private lateinit var currentUser: Notification.UserDetails

    @InjectMocks
    private lateinit var service: NotificationService

    @Mock
    private lateinit var notificationRepository: NotificationRepository

    private val modelMapper: ModelMapper = ModelMapper()

    @BeforeEach
    fun setUp() {
        currentUser = Notification.UserDetails(id = CURRENT_USER_ID, name = "From user")
        friendRequestNotification = Notification(
            notificationId = 1L,
            subjectId = TO_USER_ID,
            type = NotificationType.FRIEND_REQUEST,
            user = currentUser
        )
    }

    @Test
    fun testFindAllByUserWhenUserHasNotifications_thenReturnListOfNotificationDto() {
        val firstPage: Pageable = PageRequest.of(0, 2)
        Mockito.`when`(notificationRepository.findAllByUser_idOrderByTimestamp(TO_USER_ID, firstPage))
            .thenReturn(
                listOf(
                    Notification(user = currentUser),
                    Notification(user = currentUser)
                )
            )
        val notifications = service.findAllByUserId(TO_USER_ID, firstPage)

        Assertions.assertThat(notifications).hasSize(2)
        Assertions.assertThat(notifications[0].user!!.id).isEqualTo(currentUser.id)
    }

    @Test
    fun testUpdateNotificationToReadWhenFound_thenSetIsReadToTrue() {
        friendRequestNotification.seen = true
        val notification = friendRequestNotification.copy(
            seen = false
        )

        Mockito.`when`(notificationRepository.findById(friendRequestNotification.notificationId!!))
            .thenReturn(Optional.ofNullable(notification))

        Mockito.`when`(
            notificationRepository.save(
                Mockito.any(
                    Notification::class.java
                )
            )
        )
            .thenReturn(friendRequestNotification)

        val notificationDtoToUpdate = modelMapper.map(notification, NotificationDto::class.java)
        val updatedNotificationDto = service.updateNotificationToRead(notificationDtoToUpdate)

        Assertions.assertThat(updatedNotificationDto.seen).isTrue
    }

    @Test
    fun testUpdateNotificationToReadWhenNotFound_thenThrowEntityNotFoundException() {
        Mockito.`when`(notificationRepository.findById(friendRequestNotification.notificationId!!))
            .thenReturn(Optional.empty())

        val notificationDtoToUpdate = modelMapper.map(friendRequestNotification, NotificationDto::class.java)

        Assertions.assertThatExceptionOfType(APIException.EntityNotFoundException::class.java)
            .isThrownBy(ThrowingCallable { service.updateNotificationToRead(notificationDtoToUpdate) })
    }
}